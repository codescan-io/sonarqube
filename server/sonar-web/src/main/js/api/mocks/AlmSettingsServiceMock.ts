/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { mockAlmSettingsInstance } from '../../helpers/mocks/alm-settings';
import { AlmKeys, AlmSettingsInstance } from '../../types/alm-settings';
import { getAlmSettings } from '../alm-settings';

export default class AlmSettingsServiceMock {
  almSettings: AlmSettingsInstance[];
  defaultSetting: AlmSettingsInstance[] = [
    mockAlmSettingsInstance({ key: 'conf-final-1', alm: AlmKeys.GitLab }),
    mockAlmSettingsInstance({ key: 'conf-final-2', alm: AlmKeys.GitLab }),
    mockAlmSettingsInstance({ key: 'conf-github-1', alm: AlmKeys.GitHub, url: 'http://url' }),
    mockAlmSettingsInstance({ key: 'conf-github-2', alm: AlmKeys.GitHub, url: 'http://url' }),
    mockAlmSettingsInstance({ key: 'conf-github-3', alm: AlmKeys.GitHub, url: 'javascript://url' }),
    mockAlmSettingsInstance({ key: 'conf-azure-1', alm: AlmKeys.Azure, url: 'url' }),
    mockAlmSettingsInstance({ key: 'conf-azure-2', alm: AlmKeys.Azure, url: 'url' }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketcloud-1',
      alm: AlmKeys.BitbucketCloud,
      url: 'url',
    }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketcloud-2',
      alm: AlmKeys.BitbucketCloud,
      url: 'url',
    }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketserver-1',
      alm: AlmKeys.BitbucketServer,
      url: 'url',
    }),
    mockAlmSettingsInstance({
      key: 'conf-bitbucketserver-2',
      alm: AlmKeys.BitbucketServer,
      url: 'url',
    }),
  ];

  constructor() {
    this.almSettings = cloneDeep(this.defaultSetting);
    jest.mocked(getAlmSettings).mockImplementation(this.getAlmSettingsHandler);
  }

  getAlmSettingsHandler = () => {
    return Promise.resolve(this.almSettings);
  };

  reset = () => {
    this.almSettings = cloneDeep(this.defaultSetting);
  };
}
