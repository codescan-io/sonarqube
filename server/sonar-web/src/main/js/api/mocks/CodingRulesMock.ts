/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep, countBy, pick, trim } from 'lodash';
import { RuleDescriptionSections } from '../../apps/coding-rules/rule';
import {
  mockCurrentUser,
  mockQualityProfile,
  mockRuleDetails,
  mockRuleRepository,
} from '../../helpers/testMocks';
import { RuleRepository } from '../../types/coding-rules';
import { RawIssuesResponse } from '../../types/issues';
import { SearchRulesQuery } from '../../types/rules';
import { Rule, RuleActivation, RuleDetails, RulesUpdateRequest } from '../../types/types';
import { NoticeType } from '../../types/users';
import { getFacet } from '../issues';
import {
  bulkActivateRules,
  bulkDeactivateRules,
  Profile,
  searchQualityProfiles,
  SearchQualityProfilesParameters,
  SearchQualityProfilesResponse,
} from '../quality-profiles';
import { getRuleDetails, getRulesApp, searchRules, updateRule } from '../rules';
import { dismissNotice, getCurrentUser } from '../users';

interface FacetFilter {
  languages?: string;
}

const FACET_RULE_MAP: { [key: string]: keyof Rule } = {
  languages: 'lang',
  types: 'type',
};
export default class CodingRulesMock {
  defaultRules: RuleDetails[] = [];
  rules: RuleDetails[] = [];
  qualityProfile: Profile[] = [];
  repositories: RuleRepository[] = [];
  isAdmin = false;
  applyWithWarning = false;
  dismissedNoticesEP = false;

  constructor() {
    this.repositories = [
      mockRuleRepository({ key: 'repo1' }),
      mockRuleRepository({ key: 'repo2' }),
    ];
    this.qualityProfile = [
      mockQualityProfile({ key: 'p1', name: 'QP Foo', language: 'java', languageName: 'Java' }),
      mockQualityProfile({ key: 'p2', name: 'QP Bar', language: 'js' }),
      mockQualityProfile({ key: 'p3', name: 'QP FooBar', language: 'java', languageName: 'Java' }),
    ];

    const resourceContent = 'Some link <a href="http://example.com">Awsome Reading</a>';
    const introTitle = 'Introduction to this rule';
    const rootCauseContent = 'This how to fix';

    this.defaultRules = [
      mockRuleDetails({
        key: 'rule1',
        type: 'BUG',
        lang: 'java',
        langName: 'Java',
        name: 'Awsome java rule',
      }),
      mockRuleDetails({
        key: 'rule2',
        name: 'Hot hotspot',
        type: 'SECURITY_HOTSPOT',
        lang: 'js',
        descriptionSections: [
          { key: RuleDescriptionSections.INTRODUCTION, content: introTitle },
          { key: RuleDescriptionSections.ROOT_CAUSE, content: rootCauseContent },
          { key: RuleDescriptionSections.ASSESS_THE_PROBLEM, content: 'Assess' },
          {
            key: RuleDescriptionSections.RESOURCES,
            content: resourceContent,
          },
        ],
        langName: 'JavaScript',
      }),
      mockRuleDetails({ key: 'rule3', name: 'Unknown rule', lang: 'js', langName: 'JavaScript' }),
      mockRuleDetails({
        key: 'rule4',
        type: 'BUG',
        lang: 'c',
        langName: 'C',
        name: 'Awsome C rule',
      }),
      mockRuleDetails({
        key: 'rule5',
        type: 'VULNERABILITY',
        lang: 'py',
        langName: 'Python',
        name: 'Awsome Python rule',
        descriptionSections: [
          { key: RuleDescriptionSections.INTRODUCTION, content: introTitle },
          { key: RuleDescriptionSections.HOW_TO_FIX, content: rootCauseContent },
          {
            key: RuleDescriptionSections.RESOURCES,
            content: resourceContent,
          },
        ],
      }),
      mockRuleDetails({
        key: 'rule6',
        type: 'VULNERABILITY',
        lang: 'py',
        langName: 'Python',
        name: 'Bad Python rule',
        isExternal: true,
        descriptionSections: undefined,
      }),
      mockRuleDetails({
        key: 'rule7',
        type: 'VULNERABILITY',
        lang: 'py',
        langName: 'Python',
        name: 'Python rule with context',
        descriptionSections: [
          {
            key: RuleDescriptionSections.INTRODUCTION,
            content: 'Introduction to this rule with context',
          },
          {
            key: RuleDescriptionSections.HOW_TO_FIX,
            content: 'This is how to fix for spring',
            context: { key: 'spring', displayName: 'Spring' },
          },
          {
            key: RuleDescriptionSections.HOW_TO_FIX,
            content: 'This is how to fix for spring boot',
            context: { key: 'spring_boot', displayName: 'Spring boot' },
          },
          {
            key: RuleDescriptionSections.RESOURCES,
            content: resourceContent,
          },
        ],
      }),
      mockRuleDetails({
        key: 'rule8',
        type: 'VULNERABILITY',
        lang: 'py',
        langName: 'Python',
        name: 'Awesome Python rule with education principles',
        descriptionSections: [
          { key: RuleDescriptionSections.INTRODUCTION, content: introTitle },
          { key: RuleDescriptionSections.HOW_TO_FIX, content: rootCauseContent },
          {
            key: RuleDescriptionSections.RESOURCES,
            content: resourceContent,
          },
        ],
        educationPrinciples: ['defense_in_depth', 'never_trust_user_input'],
      }),
    ];

    (updateRule as jest.Mock).mockImplementation(this.handleUpdateRule);
    (searchRules as jest.Mock).mockImplementation(this.handleSearchRules);
    (getRuleDetails as jest.Mock).mockImplementation(this.handleGetRuleDetails);
    (searchQualityProfiles as jest.Mock).mockImplementation(this.handleSearchQualityProfiles);
    (getRulesApp as jest.Mock).mockImplementation(this.handleGetRulesApp);
    (bulkActivateRules as jest.Mock).mockImplementation(this.handleBulkActivateRules);
    (bulkDeactivateRules as jest.Mock).mockImplementation(this.handleBulkDeactivateRules);
    (getFacet as jest.Mock).mockImplementation(this.handleGetGacet);
    (getCurrentUser as jest.Mock).mockImplementation(this.handleGetCurrentUser);
    (dismissNotice as jest.Mock).mockImplementation(this.handleDismissNotification);
    this.rules = cloneDeep(this.defaultRules);
  }

  getRuleWithoutDetails() {
    return this.rules.map((r) =>
      pick(r, [
        'isTemplate',
        'key',
        'lang',
        'langName',
        'name',
        'params',
        'severity',
        'status',
        'sysTags',
        'tags',
        'type',
      ])
    );
  }

  filterFacet({ languages }: FacetFilter) {
    let filteredRules = this.getRuleWithoutDetails();
    if (languages) {
      filteredRules = filteredRules.filter((r) => r.lang && languages.includes(r.lang));
    }
    return filteredRules;
  }

  setIsAdmin() {
    this.isAdmin = true;
  }

  activateWithWarning() {
    this.applyWithWarning = true;
  }

  reset() {
    this.isAdmin = false;
    this.applyWithWarning = false;
    this.dismissedNoticesEP = false;
    this.rules = cloneDeep(this.defaultRules);
  }

  allRulesCount() {
    return this.rules.length;
  }

  allRulesName() {
    return this.rules.map((r) => r.name);
  }

  allQualityProfile(language: string) {
    return this.qualityProfile.filter((qp) => qp.language === language);
  }

  handleGetGacet = (): Promise<{
    facet: { count: number; val: string }[];
    response: RawIssuesResponse;
  }> => {
    return this.reply({
      facet: [],
      response: {
        components: [],
        effortTotal: 0,
        facets: [],
        issues: [],
        languages: [],
        paging: { total: 0, pageIndex: 1, pageSize: 1 },
      },
    });
  };

  handleGetRuleDetails = (parameters: {
    actives?: boolean;
    key: string;
  }): Promise<{ actives?: RuleActivation[]; rule: RuleDetails }> => {
    const rule = this.rules.find((r) => r.key === parameters.key);
    if (!rule) {
      return Promise.reject({
        errors: [{ msg: `No rule has been found for id ${parameters.key}` }],
      });
    }
    return this.reply({ actives: parameters.actives ? [] : undefined, rule });
  };

  handleUpdateRule = (data: RulesUpdateRequest): Promise<RuleDetails> => {
    const rule = this.rules.find((r) => r.key === data.key);
    if (rule === undefined) {
      return Promise.reject({
        errors: [{ msg: `No rule has been found for id ${data.key}` }],
      });
    }
    const template = this.rules.find((r) => r.key === rule.templateKey);

    // Lets not convert the md to html in test.
    rule.mdDesc = data.markdown_description !== undefined ? data.markdown_description : rule.mdDesc;
    rule.htmlDesc =
      data.markdown_description !== undefined ? data.markdown_description : rule.htmlDesc;
    rule.mdNote = data.markdown_note !== undefined ? data.markdown_note : rule.mdNote;
    rule.htmlNote = data.markdown_note !== undefined ? data.markdown_note : rule.htmlNote;
    rule.name = data.name !== undefined ? data.name : rule.name;
    if (template && data.params) {
      rule.params = [];
      data.params.split(';').forEach((param) => {
        const parts = param.split('=');
        const paramsDef = template.params?.find((p) => p.key === parts[0]);
        rule.params?.push({
          key: parts[0],
          type: paramsDef?.type || 'STRING',
          defaultValue: trim(parts[1], '" '),
          htmlDesc: paramsDef?.htmlDesc,
        });
      });
    }

    rule.remFnBaseEffort =
      data.remediation_fn_base_effort !== undefined
        ? data.remediation_fn_base_effort
        : rule.remFnBaseEffort;
    rule.remFnType =
      data.remediation_fn_type !== undefined ? data.remediation_fn_type : rule.remFnType;
    rule.severity = data.severity !== undefined ? data.severity : rule.severity;
    rule.status = data.status !== undefined ? data.status : rule.status;
    rule.tags = data.tags !== undefined ? data.tags.split(';') : rule.tags;

    return this.reply(rule);
  };

  handleSearchRules = ({ facets, languages, p, ps, rule_key }: SearchRulesQuery) => {
    const countFacet = (facets || '').split(',').map((facet: keyof Rule) => {
      const facetCount = countBy(
        this.rules.map((r) => r[FACET_RULE_MAP[facet] || facet] as string)
      );
      return {
        property: facet,
        values: Object.keys(facetCount).map((val) => ({ val, count: facetCount[val] })),
      };
    });
    const currentPs = ps || 10;
    const currentP = p || 1;
    let filteredRules: Rule[] = [];
    if (rule_key) {
      filteredRules = this.getRuleWithoutDetails().filter((r) => r.key === rule_key);
    } else {
      filteredRules = this.filterFacet({ languages });
    }
    const responseRules = filteredRules.slice((currentP - 1) * currentPs, currentP * currentPs);
    return this.reply({
      total: filteredRules.length,
      p: currentP,
      ps: currentPs,
      rules: responseRules,
      facets: countFacet,
    });
  };

  handleBulkActivateRules = () => {
    if (this.applyWithWarning) {
      return this.reply({
        succeeded: this.rules.length - 1,
        failed: 1,
        errors: [{ msg: 'c rule c:S6069 cannot be activated on cpp profile SonarSource' }],
      });
    }
    return this.reply({
      succeeded: this.rules.length,
      failed: 0,
      errors: [],
    });
  };

  handleBulkDeactivateRules = () => {
    return this.reply({
      succeeded: this.rules.length,
      failed: 0,
    });
  };

  handleSearchQualityProfiles = ({
    language,
  }: SearchQualityProfilesParameters = {}): Promise<SearchQualityProfilesResponse> => {
    let profiles: Profile[] = this.isAdmin
      ? this.qualityProfile.map((p) => ({ ...p, actions: { edit: true } }))
      : this.qualityProfile;
    if (language) {
      profiles = profiles.filter((p) => p.language === language);
    }
    return this.reply({ profiles });
  };

  handleGetRulesApp = () => {
    return this.reply({ canWrite: this.isAdmin, repositories: this.repositories });
  };

  handleGetCurrentUser = () => {
    return this.reply(
      mockCurrentUser({
        dismissedNotices: {
          educationPrinciples: this.dismissedNoticesEP,
        },
      })
    );
  };

  handleDismissNotification = (noticeType: NoticeType) => {
    if (noticeType === NoticeType.EDUCATION_PRINCIPLES) {
      this.dismissedNoticesEP = true;
      return this.reply(true);
    }

    return Promise.reject();
  };

  reply<T>(response: T): Promise<T> {
    return Promise.resolve(cloneDeep(response));
  }
}
