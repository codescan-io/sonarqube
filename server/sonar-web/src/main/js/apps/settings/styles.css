/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.settings-layout {
  margin-bottom: 60px;
}

.settings-definitions-list > li + li {
  margin-top: 30px;
}

.settings-definition {
  margin: 0 -20px;
  padding: 10px 20px;
  border-top: 1px solid rgba(0, 0, 0, 0);
  border-bottom: 1px solid rgba(0, 0, 0, 0);
  display: flex;
  align-items: stretch;
}

.settings-definition-changed {
  border-top: 1px solid var(--alertBorderWarning);
  border-bottom: 1px solid var(--alertBorderWarning);
  background-color: var(--alertBackgroundWarning);
}

.settings-definition-left {
  width: 330px;
  padding-right: 30px;
  box-sizing: border-box;
}

.settings-definition-right {
  position: relative;
  width: calc(100% - 330px);
  box-sizing: border-box;
}

.settings-definition-name {
  text-overflow: ellipsis;
}

.settings-definition-key {
  line-height: 1.5;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.settings-definition-key:hover {
  overflow: visible;
}

.settings-definition-state {
  min-height: 32px;
  padding-bottom: 4px;
}

.settings-definition-state > span {
  display: flex;
}

.settings-definition-changes {
  margin-top: 20px;
  padding-top: 20px;
  border-top: 1px dotted var(--barBorderColor);
}

.settings-sub-categories-list > li + li,
.settings-sub-category {
  margin: 30px -20px 0;
  padding: 30px 20px;
  border-top: 1px solid var(--barBorderColor);
}

.settings-sub-category-name {
  margin-bottom: 20px;
  font-size: var(--bigFontSize);
}

.settings-sub-category-description {
  margin-top: -15px;
  margin-bottom: 20px;
  color: var(--secondFontColor);
}

.settings-large-input {
  width: 100% !important;
  max-width: 400px;
  min-width: 200px;
}

.button {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  vertical-align: middle;
  height: 24px;
  line-height: 22px;
  padding: 0 8px;
  border: 1px solid #236a97;
  border-radius: 2px;
  box-sizing: border-box;
  background: transparent;
  color: #236a97;
  font-weight: 600;
  font-size: 12px;
  text-decoration: none;
  cursor: pointer;
  outline: none;
  transition: border-color 0.2s ease, box-shadow 0.2s ease;
}

.button:hover,
.button.button-active {
  background: #236a97;
  color: #fff;
}

.button:active {
  box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.125);
}

.button:focus {
  box-shadow: 0 0 0 3px rgba(35, 106, 151, 0.25);
}

.button-primary {
  background: #236a97;
  color: #fff;
}

.button-primary:hover {
  background: #0E516F;
}
