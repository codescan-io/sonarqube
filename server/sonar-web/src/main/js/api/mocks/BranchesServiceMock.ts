/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { mockBranch } from '../../helpers/mocks/branch-like';
import { BranchLike } from '../../types/branch-like';
import { getBranches } from '../branches';

export default class BranchesServiceMock {
  branchLikes: BranchLike[];
  defaultBranchLikes: BranchLike[] = [
    mockBranch({ isMain: true, name: 'master' }),
    mockBranch({ excludedFromPurge: false, name: 'delete-branch' }),
    mockBranch({ name: 'normal-branch' }),
  ];

  constructor() {
    this.branchLikes = cloneDeep(this.defaultBranchLikes);
    (getBranches as jest.Mock).mockImplementation(this.getBranchesHandler);
  }

  getBranchesHandler = () => {
    return Promise.resolve(this.branchLikes);
  };

  resetBranches = () => {
    this.branchLikes = cloneDeep(this.defaultBranchLikes);
  };
}
