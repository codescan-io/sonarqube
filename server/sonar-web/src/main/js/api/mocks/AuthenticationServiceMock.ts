/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { mockSettingValue } from '../../helpers/mocks/settings';
import { BranchParameters } from '../../types/branch-like';
import { SettingDefinition, SettingValue } from '../../types/settings';
import { getValues, resetSettingValue, setSettingValue } from '../settings';

export default class AuthenticationServiceMock {
  settingValues: SettingValue[];
  defaulSettingValues: SettingValue[] = [
    mockSettingValue({ key: 'test1', value: '' }),
    mockSettingValue({ key: 'test2', value: 'test2' }),
    mockSettingValue({ key: 'sonar.auth.saml.certificate.secured' }),
    mockSettingValue({ key: 'sonar.auth.saml.enabled', value: 'false' }),
  ];

  constructor() {
    this.settingValues = cloneDeep(this.defaulSettingValues);
    (getValues as jest.Mock).mockImplementation(this.getValuesHandler);
    (setSettingValue as jest.Mock).mockImplementation(this.setValueHandler);
    (resetSettingValue as jest.Mock).mockImplementation(this.resetValueHandler);
  }

  getValuesHandler = (data: { keys: string; component?: string } & BranchParameters) => {
    if (data.keys) {
      return Promise.resolve(this.settingValues.filter((set) => data.keys.includes(set.key)));
    }
    return Promise.resolve(this.settingValues);
  };

  setValueHandler = (definition: SettingDefinition, value: string) => {
    if (value === 'error') {
      const res = new Response('', {
        status: 400,
        statusText: 'fail',
      });

      return Promise.reject(res);
    }
    const updatedSettingValue = this.settingValues.find((set) => set.key === definition.key);
    if (updatedSettingValue) {
      updatedSettingValue.value = value;
    }
    return Promise.resolve();
  };

  resetValueHandler = (data: { keys: string; component?: string } & BranchParameters) => {
    if (data.keys) {
      this.settingValues.forEach((set) => {
        if (data.keys.includes(set.key)) {
          set.value = '';
        }
        return set;
      });
    }
    return Promise.resolve();
  };

  resetValues = () => {
    this.settingValues = cloneDeep(this.defaulSettingValues);
  };
}
