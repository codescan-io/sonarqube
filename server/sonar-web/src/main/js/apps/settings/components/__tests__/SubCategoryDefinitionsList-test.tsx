/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { mockSettingWithCategory } from '../../../../helpers/mocks/settings';
import { mockLocation } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import {
  SubCategoryDefinitionsList,
  SubCategoryDefinitionsListProps,
} from '../SubCategoryDefinitionsList';

jest.mock('../../../../helpers/scrolling', () => ({
  scrollToElement: jest.fn(),
}));

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('');
  expect(shallowRender({ subCategory: 'qg' })).toMatchSnapshot('subcategory');
});

it('should scroll if hash is defined and updated', async () => {
  window.HTMLElement.prototype.scrollIntoView = jest.fn();
  const wrapper = shallowRender({ location: mockLocation({ hash: '#qg' }) });

  await waitAndUpdate(wrapper);

  wrapper.find('h2').forEach((node) => mount(node.getElement()));

  expect(window.HTMLElement.prototype.scrollIntoView).toHaveBeenCalled();

  wrapper.setProps({ location: mockLocation({ hash: '#email' }) });

  expect(window.HTMLElement.prototype.scrollIntoView).toHaveBeenCalled();
});

function shallowRender(props: Partial<SubCategoryDefinitionsListProps> = {}) {
  return shallow<SubCategoryDefinitionsListProps>(
    <SubCategoryDefinitionsList
      category="general"
      location={mockLocation()}
      settings={[
        mockSettingWithCategory(),
        mockSettingWithCategory({
          definition: {
            key: 'qg',
            category: 'general',
            subCategory: 'qg',
            fields: [],
            options: [],
            description: 'awesome description',
          },
        }),
      ]}
      {...props}
    />
  );
}
