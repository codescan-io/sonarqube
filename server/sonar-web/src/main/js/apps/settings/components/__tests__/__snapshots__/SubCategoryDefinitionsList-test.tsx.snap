// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<ul
  className="settings-sub-categories-list"
>
  <li
    key="email"
  >
    <h2
      className="settings-sub-category-name"
      data-key="email"
    >
      property.category.general.email
    </h2>
    <div
      className="settings-sub-category-description markdown"
      dangerouslySetInnerHTML={
        {
          "__html": "property.category.general.email.description",
        }
      }
    />
    <DefinitionsList
      scrollToDefinition={[Function]}
      settings={
        [
          {
            "definition": {
              "category": "general",
              "description": "When Foo then Bar",
              "fields": [],
              "key": "foo",
              "name": "Foo setting",
              "options": [],
              "subCategory": "email",
              "type": "INTEGER",
            },
            "hasValue": true,
            "inherited": true,
            "key": "foo",
            "value": "42",
          },
        ]
      }
    />
    <withCurrentUserContext(EmailForm) />
  </li>
  <li
    key="qg"
  >
    <h2
      className="settings-sub-category-name"
      data-key="qg"
    >
      property.category.general.qg
    </h2>
    <div
      className="settings-sub-category-description markdown"
      dangerouslySetInnerHTML={
        {
          "__html": "property.category.general.qg.description",
        }
      }
    />
    <DefinitionsList
      scrollToDefinition={[Function]}
      settings={
        [
          {
            "definition": {
              "category": "general",
              "description": "awesome description",
              "fields": [],
              "key": "qg",
              "options": [],
              "subCategory": "qg",
            },
            "hasValue": true,
            "inherited": true,
            "key": "foo",
            "value": "42",
          },
        ]
      }
    />
  </li>
</ul>
`;

exports[`should render correctly: subcategory 1`] = `
<ul
  className="settings-sub-categories-list"
>
  <li
    key="qg"
  >
    <h2
      className="settings-sub-category-name"
      data-key="qg"
    >
      property.category.general.qg
    </h2>
    <div
      className="settings-sub-category-description markdown"
      dangerouslySetInnerHTML={
        {
          "__html": "property.category.general.qg.description",
        }
      }
    />
    <DefinitionsList
      scrollToDefinition={[Function]}
      settings={
        [
          {
            "definition": {
              "category": "general",
              "description": "awesome description",
              "fields": [],
              "key": "qg",
              "options": [],
              "subCategory": "qg",
            },
            "hasValue": true,
            "inherited": true,
            "key": "foo",
            "value": "42",
          },
        ]
      }
    />
  </li>
</ul>
`;
